### ISSUE TITLE
failed to open CUDA IPC handle at cuda initialization

### ISSUE THREAD

**Issue by WencongXiao** [NVIDIA/nccl#360]
Hi, 

I run pytorch with the latest nccl release 2.7.8 in a 8 V100 GPU server trying to use NVLINK for DDP communication.

Different from the traditional approach of using NVLINK in the same container, I try to run each python process in a docker container, with all GPUs visible to use NVLINK for communication across containers.  
Actually, sometimes it can success with 2 workers (each runs with a GPU) or 3 workers or 4 workers. However sometimes it will randomly fail when doing nccl initialization, showing a non-deterministic behavior. Note that, I find that the success ratio drops with more workers. I never run it successfully with 8 workers.

The failed logs are listed as below. 
```sh
iZ8vb69v6s60nfnp50dd1dZ:44:44 [7] NCCL INFO Bootstrap : Using [0]eth0:172.26.10.27<0>
iZ8vb69v6s60nfnp50dd1dZ:44:44 [7] NCCL INFO NET/Plugin : No plugin found (libnccl-net.so), using internal implementation
iZ8vb69v6s60nfnp50dd1dZ:44:44 [7] NCCL INFO NET/IB : No device found.
iZ8vb69v6s60nfnp50dd1dZ:44:44 [7] NCCL INFO NET/Socket : Using [0]eth0:172.26.10.27<0>
iZ8vb69v6s60nfnp50dd1dZ:44:44 [7] NCCL INFO Using network Socket
iZ8vb69v6s60nfnp50dd1dZ:44:50 [7] NCCL INFO NCCL_MAX_NRINGS set by environment to 4.
iZ8vb69v6s60nfnp50dd1dZ:44:50 [7] NCCL INFO NCCL_MAX_NCHANNELS set by environment to 4.
iZ8vb69v6s60nfnp50dd1dZ:44:50 [7] NCCL INFO NCCL_MIN_NRINGS set by environment to 4.
iZ8vb69v6s60nfnp50dd1dZ:44:50 [7] NCCL INFO NCCL_MIN_NCHANNELS set by environment to 4.
iZ8vb69v6s60nfnp50dd1dZ:44:50 [7] NCCL INFO threadThresholds 8/8/64 | 64/8/64 | 8/8/64
iZ8vb69v6s60nfnp50dd1dZ:44:50 [7] NCCL INFO Trees [0] 4/-1/-1->7->1|1->7->4/-1/-1 [1] 4/-1/-1->7->1|1->7->4/-1/-1 [2] 1/-1/-1->7->4|4->7->1/-1/-1 [3] 1/-1/-1->7->4|4->7->1/-1/-1
iZ8vb69v6s60nfnp50dd1dZ:44:50 [7] NCCL INFO Channel 00 : 7[e0] -> 4[b0] via P2P/IPC

iZ8vb69v6s60nfnp50dd1dZ:44:50 [7] transport/p2p.cc:238 NCCL WARN failed to open CUDA IPC handle : 10 invalid device ordinal
iZ8vb69v6s60nfnp50dd1dZ:44:50 [7] NCCL INFO transport.cc:68 -> 1
iZ8vb69v6s60nfnp50dd1dZ:44:50 [7] NCCL INFO init.cc:766 -> 1
iZ8vb69v6s60nfnp50dd1dZ:44:50 [7] NCCL INFO init.cc:840 -> 1
iZ8vb69v6s60nfnp50dd1dZ:44:50 [7] NCCL INFO group.cc:73 -> 1 [Async thread]
Traceback (most recent call last):
  File "mnist.py", line 195, in <module>
    main()
  File "mnist.py", line 181, in main
    model = torch.nn.parallel.DistributedDataParallel(model, device_ids=[gpu_id], output_device=gpu_id)
  File "/home/pai/lib/python3.7/site-packages/torch/nn/parallel/distributed.py", line 298, in __init__
    self.broadcast_bucket_size)
  File "/home/pai/lib/python3.7/site-packages/torch/nn/parallel/distributed.py", line 480, in _distributed_broadcast_coalesced
    dist._broadcast_coalesced(self.process_group, tensors, buffer_size)
RuntimeError: NCCL error in: /workspace/pytorch/torch/lib/c10d/ProcessGroupNCCL.cpp:400, unhandled cuda error
```
It seems to me that ``transport/p2p.cc:238 NCCL WARN failed to open CUDA IPC handle : 10 invalid device ordinal`` is the root cause, as it will not appear in the successful cases.

To help you reproduce my case, I include all the necessnary information below. Please let me know for any problems.

Docker version & nvidia-driver information
```sh
[root@iZ8vb69v6s60nfnp50dd1dZ ~]# docker --version
Docker version 19.03.12, build 48a66213fe
[root@iZ8vb69v6s60nfnp50dd1dZ ~]# nvidia-smi
Fri Jul 31 10:57:38 2020
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 418.126.02   Driver Version: 418.126.02   CUDA Version: 10.1     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla V100-SXM2...  On   | 00000000:00:07.0 Off |                    0 |
| N/A   36C    P0    41W / 300W |      0MiB / 32480MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   1  Tesla V100-SXM2...  On   | 00000000:00:08.0 Off |                    0 |
| N/A   37C    P0    40W / 300W |      0MiB / 32480MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   2  Tesla V100-SXM2...  On   | 00000000:00:09.0 Off |                    0 |
| N/A   35C    P0    42W / 300W |      0MiB / 32480MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   3  Tesla V100-SXM2...  On   | 00000000:00:0A.0 Off |                    0 |
| N/A   35C    P0    42W / 300W |      0MiB / 32480MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   4  Tesla V100-SXM2...  On   | 00000000:00:0B.0 Off |                    0 |
| N/A   36C    P0    41W / 300W |      0MiB / 32480MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   5  Tesla V100-SXM2...  On   | 00000000:00:0C.0 Off |                    0 |
| N/A   35C    P0    41W / 300W |      0MiB / 32480MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   6  Tesla V100-SXM2...  On   | 00000000:00:0D.0 Off |                    0 |
| N/A   34C    P0    43W / 300W |      0MiB / 32480MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   7  Tesla V100-SXM2...  On   | 00000000:00:0E.0 Off |                    0 |
| N/A   37C    P0    40W / 300W |      0MiB / 32480MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+
```

Here's the docker launch script.
```sh
num=8
for((i=0;i<${num};i++));
do
  echo ${i}
  docker run --gpus all -tid --net=host --ipc=host --name=w${i} -v /root:/root registry.cn-zhangjiakou.aliyuncs.com/xiaowencong/pai-dlc-test:torch-training-1.3-gpu-py3-nccl-2.7.8 bash
done
```

Here's the pytorch execution script.
```sh
num=8
for((i=0;i<${num};i++));
do
  echo ${i}
  docker exec w${i} bash -c "cd /root && MASTER_ADDR=127.0.0.1 MASTER_PORT=56789 WORLD_SIZE=${num} RANK=${i} VISIBLE_DEVICE_LIST=${i} python mnist.py &> ${i}.txt" &
  sleep 1
done
```

Here's the pytorch code, a simple mnist benchmark. For the first time to run it, you might need to run just 1 worker to download data into the ./data folder. 
```python
from __future__ import print_function
import argparse
import math
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from torch.utils.data.sampler import Sampler
from torchvision import datasets, transforms
from torch.optim.lr_scheduler import StepLR

import os
os.environ["NCCL_DEBUG"]="INFO"

import torch.distributed as dist
gpu_id = int(os.environ.get("VISIBLE_DEVICE_LIST"))
torch.cuda.set_device(gpu_id)
dist.init_process_group("nccl")

class DistributedSampler(Sampler):
  """Sampler that restricts data loading to a subset of the dataset.
  It is especially useful in conjunction with
  :class:`torch.nn.parallel.DistributedDataParallel`. In such case, each
  process can pass a DistributedSampler instance as a DataLoader sampler,
  and load a subset of the original dataset that is exclusive to it.
  .. note::
      Dataset is assumed to be of constant size.
  Arguments:
      dataset: Dataset used for sampling.
      num_replicas (optional): Number of processes participating in
          distributed training.
      rank (optional): Rank of the current process within num_replicas.
  """

  def __init__(self, dataset, num_replicas=None, rank=None):
    if num_replicas is None:
      num_replicas = dist.get_world_size()
    if rank is None:
      rank = dist.get_rank()
    self.dataset = dataset
    self.num_replicas = num_replicas
    self.rank = rank
    self.epoch = 0
    self.num_samples = int(math.ceil(len(self.dataset) * 1.0 / self.num_replicas))
    self.total_size = self.num_samples * self.num_replicas

  def __iter__(self):
    # deterministically shuffle based on epoch
    g = torch.Generator()
    g.manual_seed(self.epoch)
    indices = list(torch.randperm(len(self.dataset), generator=g))

    # add extra samples to make it evenly divisible
    indices += indices[:(self.total_size - len(indices))]
    assert len(indices) == self.total_size

    # subsample
    offset = self.num_samples * self.rank
    indices = indices[offset:offset + self.num_samples]
    assert len(indices) == self.num_samples

    return iter(indices)

  def __len__(self):
    return self.num_samples

  def set_epoch(self, epoch):
    self.epoch = epoch

class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = nn.Conv2d(1, 32, 3, 1)
        self.conv2 = nn.Conv2d(32, 64, 3, 1)
        self.dropout1 = nn.Dropout2d(0.25)
        self.dropout2 = nn.Dropout2d(0.5)
        self.fc1 = nn.Linear(9216, 128)
        self.fc2 = nn.Linear(128, 10)

    def forward(self, x):
        x = self.conv1(x)
        x = F.relu(x)
        x = self.conv2(x)
        x = F.relu(x)
        x = F.max_pool2d(x, 2)
        x = self.dropout1(x)
        x = torch.flatten(x, 1)
        x = self.fc1(x)
        x = F.relu(x)
        x = self.dropout2(x)
        x = self.fc2(x)
        output = F.log_softmax(x, dim=1)
        return output


def train(args, model, device, train_loader, optimizer, epoch):
    model.train()
    for batch_idx, (data, target) in enumerate(train_loader):
        data, target = data.to(device), target.to(device)
        optimizer.zero_grad()
        output = model(data)
        loss = F.nll_loss(output, target)
        loss.backward()
        optimizer.step()
        if batch_idx % args.log_interval == 0:
            print('Train Epoch: {} [{}/{} ({:.0f}%)]\tLoss: {:.6f}'.format(
                epoch, batch_idx * len(data), len(train_loader.dataset),
                100. * batch_idx / len(train_loader), loss.item()))
            if args.dry_run:
                break


def test(model, device, test_loader):
    model.eval()
    test_loss = 0
    correct = 0
    with torch.no_grad():
        for data, target in test_loader:
            data, target = data.to(device), target.to(device)
            output = model(data)
            test_loss += F.nll_loss(output, target, reduction='sum').item()  # sum up batch loss
            pred = output.argmax(dim=1, keepdim=True)  # get the index of the max log-probability
            correct += pred.eq(target.view_as(pred)).sum().item()

    test_loss /= len(test_loader.dataset)

    print('\nTest set: Average loss: {:.4f}, Accuracy: {}/{} ({:.0f}%)\n'.format(
        test_loss, correct, len(test_loader.dataset),
        100. * correct / len(test_loader.dataset)))


def main():
    # Training settings
    parser = argparse.ArgumentParser(description='PyTorch MNIST Example')
    parser.add_argument('--batch-size', type=int, default=64, metavar='N',
                        help='input batch size for training (default: 64)')
    parser.add_argument('--test-batch-size', type=int, default=1000, metavar='N',
                        help='input batch size for testing (default: 1000)')
    parser.add_argument('--epochs', type=int, default=2, metavar='N',
                        help='number of epochs to train (default: 14)')
    parser.add_argument('--lr', type=float, default=1.0, metavar='LR',
                        help='learning rate (default: 1.0)')
    parser.add_argument('--gamma', type=float, default=0.7, metavar='M',
                        help='Learning rate step gamma (default: 0.7)')
    parser.add_argument('--no-cuda', action='store_true', default=False,
                        help='disables CUDA training')
    parser.add_argument('--dry-run', action='store_true', default=False,
                        help='quickly check a single pass')
    parser.add_argument('--seed', type=int, default=1, metavar='S',
                        help='random seed (default: 1)')
    parser.add_argument('--log-interval', type=int, default=10, metavar='N',
                        help='how many batches to wait before logging training status')
    parser.add_argument('--save-model', action='store_true', default=False,
                        help='For Saving the current Model')
    args = parser.parse_args()
    use_cuda = not args.no_cuda and torch.cuda.is_available()

    torch.manual_seed(args.seed)

    device = torch.device("cuda" if use_cuda else "cpu")

    kwargs = {'batch_size': args.batch_size}
    if use_cuda:
        kwargs.update({'num_workers': int(os.environ.get("WORLD_SIZE", 1)),
                       'pin_memory': True}
                     )

    transform=transforms.Compose([
        transforms.ToTensor(),
        transforms.Normalize((0.1307,), (0.3081,))
        ])
    dataset1 = datasets.MNIST('./data', train=True, download=True,
                       transform=transform)
    dataset2 = datasets.MNIST('./data', train=False,
                       transform=transform)
    dataset1_sampler = DistributedSampler(dataset1)
    train_loader = torch.utils.data.DataLoader(dataset1, sampler=dataset1_sampler,**kwargs)
    test_loader = torch.utils.data.DataLoader(dataset2, **kwargs)

    model = Net().cuda()
    model = torch.nn.parallel.DistributedDataParallel(model, device_ids=[gpu_id], output_device=gpu_id)
    optimizer = optim.Adadelta(model.parameters(), lr=args.lr)

    scheduler = StepLR(optimizer, step_size=1, gamma=args.gamma)
    for epoch in range(1, args.epochs + 1):
        train(args, model, device, train_loader, optimizer, epoch)
        test(model, device, test_loader)
        scheduler.step()

    if args.save_model:
        torch.save(model.state_dict(), "mnist_cnn.pt")
if __name__ == '__main__':
    main()
```

Here's the log of 2 workers successfully run with NVLINK.
```sh
iZ8vb69v6s60nfnp50dd1dZ:297:297 [0] NCCL INFO Bootstrap : Using [0]eth0:172.26.10.27<0>
iZ8vb69v6s60nfnp50dd1dZ:297:297 [0] NCCL INFO NET/Plugin : No plugin found (libnccl-net.so), using internal implementation
iZ8vb69v6s60nfnp50dd1dZ:297:297 [0] NCCL INFO NET/IB : No device found.
iZ8vb69v6s60nfnp50dd1dZ:297:297 [0] NCCL INFO NET/Socket : Using [0]eth0:172.26.10.27<0>
iZ8vb69v6s60nfnp50dd1dZ:297:297 [0] NCCL INFO Using network Socket
NCCL version 2.7.8+cuda10.0
iZ8vb69v6s60nfnp50dd1dZ:297:305 [0] NCCL INFO NCCL_MAX_NRINGS set by environment to 4.
iZ8vb69v6s60nfnp50dd1dZ:297:305 [0] NCCL INFO NCCL_MAX_NCHANNELS set by environment to 4.
iZ8vb69v6s60nfnp50dd1dZ:297:305 [0] NCCL INFO NCCL_MIN_NRINGS set by environment to 4.
iZ8vb69v6s60nfnp50dd1dZ:297:305 [0] NCCL INFO NCCL_MIN_NCHANNELS set by environment to 4.
iZ8vb69v6s60nfnp50dd1dZ:297:305 [0] NCCL INFO Channel 00/04 :    0   1
iZ8vb69v6s60nfnp50dd1dZ:297:305 [0] NCCL INFO Channel 01/04 :    0   1
iZ8vb69v6s60nfnp50dd1dZ:297:305 [0] NCCL INFO Channel 02/04 :    0   1
iZ8vb69v6s60nfnp50dd1dZ:297:305 [0] NCCL INFO Channel 03/04 :    0   1
iZ8vb69v6s60nfnp50dd1dZ:297:305 [0] NCCL INFO threadThresholds 8/8/64 | 16/8/64 | 8/8/64
iZ8vb69v6s60nfnp50dd1dZ:297:305 [0] NCCL INFO Trees [0] 1/-1/-1->0->-1|-1->0->1/-1/-1 [1] 1/-1/-1->0->-1|-1->0->1/-1/-1 [2] 1/-1/-1->0->-1|-1->0->1/-1/-1 [3] 1/-1/-1->0->-1|-1->0->1/-1/-1
iZ8vb69v6s60nfnp50dd1dZ:297:305 [0] NCCL INFO Channel 00 : 0[70] -> 1[80] via P2P/IPC
iZ8vb69v6s60nfnp50dd1dZ:297:305 [0] NCCL INFO Channel 01 : 0[70] -> 1[80] via P2P/IPC
iZ8vb69v6s60nfnp50dd1dZ:297:305 [0] NCCL INFO Channel 02 : 0[70] -> 1[80] via P2P/IPC
iZ8vb69v6s60nfnp50dd1dZ:297:305 [0] NCCL INFO Channel 03 : 0[70] -> 1[80] via P2P/IPC
iZ8vb69v6s60nfnp50dd1dZ:297:305 [0] NCCL INFO 4 coll channels, 4 p2p channels, 2 p2p channels per peer
iZ8vb69v6s60nfnp50dd1dZ:297:305 [0] NCCL INFO comm 0x7fb758000dc0 rank 0 nranks 2 cudaDev 0 busId 70 - Init COMPLETE
iZ8vb69v6s60nfnp50dd1dZ:297:297 [0] NCCL INFO Launch mode Parallel
Train Epoch: 1 [0/60000 (0%)]   Loss: 2.346260
Train Epoch: 1 [640/60000 (2%)] Loss: 1.125059
Train Epoch: 1 [1280/60000 (4%)]        Loss: 0.420499
Train Epoch: 1 [1920/60000 (6%)]        Loss: 0.435044
Train Epoch: 1 [2560/60000 (9%)]        Loss: 0.237905
Train Epoch: 1 [3200/60000 (11%)]       Loss: 0.341692
Train Epoch: 1 [3840/60000 (13%)]       Loss: 0.247542
Train Epoch: 1 [4480/60000 (15%)]       Loss: 0.281129
Train Epoch: 1 [5120/60000 (17%)]       Loss: 0.168577
Train Epoch: 1 [5760/60000 (19%)]       Loss: 0.074892
Train Epoch: 1 [6400/60000 (21%)]       Loss: 0.529991
Train Epoch: 1 [7040/60000 (23%)]       Loss: 0.123493
Train Epoch: 1 [7680/60000 (26%)]       Loss: 0.108398
Train Epoch: 1 [8320/60000 (28%)]       Loss: 0.140637
Train Epoch: 1 [8960/60000 (30%)]       Loss: 0.214743
Train Epoch: 1 [9600/60000 (32%)]       Loss: 0.255515
Train Epoch: 1 [10240/60000 (34%)]      Loss: 0.164548
Train Epoch: 1 [10880/60000 (36%)]      Loss: 0.146858
Train Epoch: 1 [11520/60000 (38%)]      Loss: 0.053951
Train Epoch: 1 [12160/60000 (41%)]      Loss: 0.117583
Train Epoch: 1 [12800/60000 (43%)]      Loss: 0.117814
Train Epoch: 1 [13440/60000 (45%)]      Loss: 0.101052
Train Epoch: 1 [14080/60000 (47%)]      Loss: 0.074723
Train Epoch: 1 [14720/60000 (49%)]      Loss: 0.112131
Train Epoch: 1 [15360/60000 (51%)]      Loss: 0.119227
Train Epoch: 1 [16000/60000 (53%)]      Loss: 0.117843
Train Epoch: 1 [16640/60000 (55%)]      Loss: 0.168275
Train Epoch: 1 [17280/60000 (58%)]      Loss: 0.116297
Train Epoch: 1 [17920/60000 (60%)]      Loss: 0.039182
Train Epoch: 1 [18560/60000 (62%)]      Loss: 0.079580
Train Epoch: 1 [19200/60000 (64%)]      Loss: 0.152522
Train Epoch: 1 [19840/60000 (66%)]      Loss: 0.159577
Train Epoch: 1 [20480/60000 (68%)]      Loss: 0.304155
Train Epoch: 1 [21120/60000 (70%)]      Loss: 0.083275
Train Epoch: 1 [21760/60000 (72%)]      Loss: 0.040437
Train Epoch: 1 [22400/60000 (75%)]      Loss: 0.144258
Train Epoch: 1 [23040/60000 (77%)]      Loss: 0.166302
Train Epoch: 1 [23680/60000 (79%)]      Loss: 0.177745
Train Epoch: 1 [24320/60000 (81%)]      Loss: 0.061851
Train Epoch: 1 [24960/60000 (83%)]      Loss: 0.048526
Train Epoch: 1 [25600/60000 (85%)]      Loss: 0.240470
Train Epoch: 1 [26240/60000 (87%)]      Loss: 0.161989
Train Epoch: 1 [26880/60000 (90%)]      Loss: 0.011752
Train Epoch: 1 [27520/60000 (92%)]      Loss: 0.138191
Train Epoch: 1 [28160/60000 (94%)]      Loss: 0.207724
Train Epoch: 1 [28800/60000 (96%)]      Loss: 0.144638
Train Epoch: 1 [29440/60000 (98%)]      Loss: 0.163271

Test set: Average loss: 0.0606, Accuracy: 9799/10000 (98%)
```

**Comment by WencongXiao** [comment]
@sjeaugey could you help? many thanks for any clues!

**Comment by kwen2501** [comment]
> I try to run each python process in a docker container, with all GPUs visible to use NVLINK for communication across containers.
Hi, have you tried using NCCL + CUDA 10.1 or later and having one GPU visible in one container only?

**Comment by WencongXiao** [comment]
> Hi, have you tried using NCCL + CUDA 10.1 or later and having one GPU visible in one container only?

Not yet. But I can have a try. 
The case mentioned above is with CUDA 10.0. That's why we put all gpus visible to containers.

**Comment by AddyLaddy** [comment]
Ok thanks for this report, we have been able to reproduce it internally on our DGX1[V] systems using CUDA 10.0 and 10.1 with the R440 device driver.
We believe the issue is related to how the CUDA IPC code performs its sanity checking on the CUDA IPC handles that NCCL creates and exchanges. In particular it doesn't allow a Linux process (PID) to open a handle back to itself. With the docker namespace isolation the PIDs in each container can be identical and hence the CUDA sanity checks fail.
I have found that adding `--pid=host` to the docker run command line resolves the NCCL CUDA IPC issue.
I will raise an internal bug against the CUDA team for them to improve the sanity checking in containerized environments.
